src/frontend/next/src/components/UserProvider.tsx

```diff
@@ -16,4 +18,63 @@ const UserProvider = ({ children }: Props) => {
 
 export default UserProvider;
 
-export const useUser = () => useContext(UserContext);
+/**
+ * User Hooks
+ * ----------
+ *
+ * We provide three different user hooks, which work in a similar way,
+ * but are meant to be used in different circumstances.
+ *
+ * useUser() - when all you care about is if the user is logged in or not.
+ *
+ * useAuthenticatedUser(shouldRedirect) - when you need an authenticated user.
+ * You will get null or an AuthenticatedUser in response.  If shouldRedirect
+ * is true, it will redirect to the Telescope login page.
+ *
+ * useAdminUser(shouldRedirect) - when you need an authenticated admin user.
+ * You will get null or an AdminUser in response.  If shouldRedirect
+ * is true, it will redirect to the Telescope login page.
+ */
+export const useUser = (): null | User => useContext(UserContext);
+
+export const useAuthenticatedUser = (shouldRedirect = false): null | AuthenticatedUser => {
+  const user = useUser();
+
+  // Wait until we have user info from the server
+  if (!user) {
+    return null;
+  }
+
+  if (!user.isLoggedIn) {
+    if (shouldRedirect) {
+      window.location.href = loginUrl;
+    }
+    // Caller wants an authenticated user, which we don't have,
+    // so return nothing.
+    return null;
+  }
+
+  // User's authentication status is known, return that
+  return user as AuthenticatedUser;
+};
+
+export const useAdminUser = (shouldRedirect = false): null | AdminUser => {
+  const user = useAuthenticatedUser();
+
+  // Wait until we have user info from the server
+  if (!user) {
+    return null;
+  }
+
+  if (!user.isAdmin) {
+    if (shouldRedirect) {
+      window.location.href = loginUrl;
+    }
+    // Caller wants an admin user, which we don't have,
+    // so return nothing.
+    return null;
+  }
+
+  // User's authentication status is known, return that
+  return user as AdminUser;
```

Comment:
Since an AdminUser is basically an Authenticated user with an isAdmin value of true, wouldn't it make sense to just have one function but check that isAdmin value? You could remove the duplicate function by just adding an AdminUser to the list of return types here.

Reply:
I could.  The reason I thought to split it out into a 3rd function is that it lets you signal intent in the components that use it:  Consider:

```
// if we only have useAuthenticatedUser()
const user = useAuthenticatedUser();  // you can't tell what I want here
if(user?.isAdmin) {
  // user is admin
}

// if we also have useAdminUser()
const user = useAdminUser; // indicating that I want an admin
if(user) {
  // user is admin
}
```